File: src/hexa_kinematics/test/test_leg_ik.py
import math

import pytest

from hexa_kinematics import (
    LegSpec,
    UnreachableTarget,
    forward_kinematics,
    inverse_kinematics,
)


LEG = LegSpec(
    mount_xyz=(0.0, 0.0, 0.0),
    mount_yaw=0.0,
    coxa_len=0.05,
    femur_len=0.08,
    tibia_len=0.12,
)


def _close(a, b, tol=1e-9):
    return all(abs(x - y) < tol for x, y in zip(a, b))


def test_fk_zero_angles_extends_leg_along_x():
    foot = forward_kinematics((0.0, 0.0, 0.0), LEG)
    assert _close(foot, (LEG.coxa_len + LEG.femur_len + LEG.tibia_len, 0.0, 0.0))


def test_ik_extended_pose_returns_zero_angles():
    target = (LEG.coxa_len + LEG.femur_len + LEG.tibia_len, 0.0, 0.0)
    angles = inverse_kinematics(target, LEG)
    assert _close(angles, (0.0, 0.0, 0.0))


def test_ik_foot_straight_down_from_femur_joint():
    target = (LEG.coxa_len, 0.0, -(LEG.femur_len + LEG.tibia_len))
    angles = inverse_kinematics(target, LEG)
    assert _close(angles, (0.0, math.pi / 2, 0.0))


def test_ik_coxa_yaw_to_the_side():
    target = (0.0, LEG.coxa_len + LEG.femur_len + LEG.tibia_len, 0.0)
    angles = inverse_kinematics(target, LEG)
    assert _close(angles, (math.pi / 2, 0.0, 0.0))


@pytest.mark.parametrize(
    "angles",
    [
        (0.0, 0.0, 0.0),
        (0.0, 0.3, 0.4),
        (0.5, 0.2, 0.6),
        (-0.5, 0.4, 0.5),
        (1.2, -0.3, 1.0),
        (-1.0, 0.7, 0.3),
        (0.0, math.pi / 4, math.pi / 3),
    ],
)
def test_fk_ik_round_trip(angles):
    foot = forward_kinematics(angles, LEG)
    recovered = inverse_kinematics(foot, LEG)
    foot_again = forward_kinematics(recovered, LEG)
    assert _close(foot, foot_again, tol=1e-9)


def test_ik_raises_on_target_beyond_reach():
    too_far = (10.0, 0.0, 0.0)
    with pytest.raises(UnreachableTarget):
        inverse_kinematics(too_far, LEG)


def test_ik_raises_on_target_inside_inner_annulus():
    # Foot at the femur joint → d = 0, below |femur - tibia|.
    too_close = (LEG.coxa_len, 0.0, 0.0)
    with pytest.raises(UnreachableTarget):
        inverse_kinematics(too_close, LEG)

File: src/hexa_kinematics_cpp/test/test_leg_ik.cpp
// Port of hexa_kinematics/test/test_leg_ik.py.
#include <gtest/gtest.h>

#include <cmath>
#include <vector>

#include "hexa_kinematics_cpp/leg_geometry.hpp"
#include "hexa_kinematics_cpp/leg_ik.hpp"

namespace {

using hexa_kinematics::JointAngles;
using hexa_kinematics::LegSpec;
using hexa_kinematics::Point3;

const LegSpec kLeg{Point3(0.0, 0.0, 0.0), 0.0, 0.05, 0.08, 0.12};

void expectClosePoint(const Point3& a, const Point3& b, double tol = 1e-9) {
  EXPECT_NEAR(a[0], b[0], tol);
  EXPECT_NEAR(a[1], b[1], tol);
  EXPECT_NEAR(a[2], b[2], tol);
}

void expectCloseAngles(const JointAngles& a, const JointAngles& b,
                       double tol = 1e-9) {
  EXPECT_NEAR(a[0], b[0], tol);
  EXPECT_NEAR(a[1], b[1], tol);
  EXPECT_NEAR(a[2], b[2], tol);
}

TEST(LegIk, FkZeroAnglesExtendsLegAlongX) {
  const Point3 foot = hexa_kinematics::forward_kinematics({0.0, 0.0, 0.0}, kLeg);
  expectClosePoint(
      foot, Point3(kLeg.coxa_len + kLeg.femur_len + kLeg.tibia_len, 0.0, 0.0));
}

TEST(LegIk, IkExtendedPoseReturnsZeroAngles) {
  const Point3 target(kLeg.coxa_len + kLeg.femur_len + kLeg.tibia_len, 0.0, 0.0);
  expectCloseAngles(hexa_kinematics::inverse_kinematics(target, kLeg),
                    {0.0, 0.0, 0.0});
}

TEST(LegIk, IkFootStraightDownFromFemurJoint) {
  const Point3 target(kLeg.coxa_len, 0.0,
                      -(kLeg.femur_len + kLeg.tibia_len));
  expectCloseAngles(hexa_kinematics::inverse_kinematics(target, kLeg),
                    {0.0, M_PI / 2.0, 0.0});
}

TEST(LegIk, IkCoxaYawToTheSide) {
  const Point3 target(0.0, kLeg.coxa_len + kLeg.femur_len + kLeg.tibia_len, 0.0);
  expectCloseAngles(hexa_kinematics::inverse_kinematics(target, kLeg),
                    {M_PI / 2.0, 0.0, 0.0});
}

TEST(LegIk, FkIkRoundTrip) {
  const std::vector<JointAngles> cases = {
      {0.0, 0.0, 0.0},
      {0.0, 0.3, 0.4},
      {0.5, 0.2, 0.6},
      {-0.5, 0.4, 0.5},
      {1.2, -0.3, 1.0},
      {-1.0, 0.7, 0.3},
      {0.0, M_PI / 4.0, M_PI / 3.0},
  };
  for (const auto& angles : cases) {
    const Point3 foot = hexa_kinematics::forward_kinematics(angles, kLeg);
    const JointAngles recovered =
        hexa_kinematics::inverse_kinematics(foot, kLeg);
    const Point3 foot_again =
        hexa_kinematics::forward_kinematics(recovered, kLeg);
    expectClosePoint(foot, foot_again, 1e-9);
  }
}

TEST(LegIk, IkRaisesOnTargetBeyondReach) {
  EXPECT_THROW(hexa_kinematics::inverse_kinematics(Point3(10.0, 0.0, 0.0), kLeg),
               hexa_kinematics::UnreachableTarget);
}

TEST(LegIk, IkRaisesOnTargetInsideInnerAnnulus) {
  // Foot at the femur joint -> d = 0, below |femur - tibia|.
  EXPECT_THROW(
      hexa_kinematics::inverse_kinematics(Point3(kLeg.coxa_len, 0.0, 0.0), kLeg),
      hexa_kinematics::UnreachableTarget);
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
